server/src/editor.js

```diff
@@ -0,0 +1,140 @@
+          <canvas className="highlighter centered" id="highlighter" ref="highlighter" height={canvasHeight} width={canvasWidth}></canvas>
+          <canvas className="editor centered" id="editor" ref="editor" height={canvasHeight} width={canvasWidth}></canvas>
+        </div>
+      </div>
+    </div>
+  }
+
+  componentDidUpdate() {
+    this.edit();
+  }
+
+  onClickCancel() {
+    this.props.onCancelEdit(false);
+    sendEvent("cancel", "annotation-toolbar");
+  }
+
+  onClickSave() {
+    sendEvent("save", "annotation-toolbar");
+    this.imageContext.drawImage(editor, 0, 0);
+    this.imageContext.globalCompositeOperation = 'multiply';
+    this.imageContext.drawImage(highlighter, 0, 0);
+    let dataUrl = this.imageCanvas.toDataURL();
+    this.props.onClickSave(dataUrl);
+  }
+
+  onClickHighlight() {
+    if (this.state.tool != 'highlighter') {
+      this.setState({tool: 'highlighter', highlighter: 'active', pen: 'inactive'});
+      sendEvent("highlighter-select", "annotation-toolbar");
+    } else {
+      this.setState({tool: 'none', highlighter: 'inactive'});
+      sendEvent("highlighter-deselect", "annotation-toolbar");
+    }
+  }
+
+  onClickPen() {
+    if (this.state.tool != 'pen') {
+      this.setState({tool: 'pen', pen: 'active', highlighter: 'inactive'});
+      sendEvent("pen-select", "annotation-toolbar");
+    } else {
+      this.setState({tool: 'none', pen: 'inactive'});
+      sendEvent("pen-deselect", "annotation-toolbar");
+    }
+  }
+
+  componentDidMount() {
+    editor = this.refs.editor;
+    context = editor.getContext('2d');
+    highlighter = this.refs.highlighter;
+    highlightContext = highlighter.getContext('2d');
+    let imageCanvas = this.refs.image;
+    this.imageCanvas = imageCanvas
+    let imageContext = imageCanvas.getContext('2d');
+    this.imageContext = imageContext;
+    // From https://blog.headspin.com/?p=464, we oversample the canvas for improved image quality
+    let img = new Image();
+    img.src = this.props.clip.image.url;
+    let width = this.props.clip.image.dimensions.x;
+    let height = this.props.clip.image.dimensions.y;
+    this.imageContext.scale(2, 2);
+    this.imageContext.drawImage(img, 0, 0, width, height);
+  }
+
+  edit() {
+    pos = { x: 0, y: 0 };
+    if (this.state.tool == 'highlighter') {
+      highlightContext.lineWidth = 20;
+      highlightContext.strokeStyle = '#ff0';
+      drawContext = highlightContext;
+    } else if (this.state.tool == 'pen') {
+      context.strokeStyle = this.state.color;
+      highlightContext.strokeStyle = this.state.color;
+      context.globalCompositeOperation = 'source-over';
+      drawContext = context;
+    }
+    context.lineWidth = this.state.size;
+    document.querySelector("#canvas-container").addEventListener("mousemove", this.draw);
```

Comment:
There needs to be a `this.draw.bind(this)` here, a function without a bind will lose its association with `this`. (Sometimes React fixes this, but not always.)  *Actually*, in `constructor()` there should probably be something like `this.draw = this.draw.bind(this)`, because function identity is going to be important for removing these functions.

Generally there needs to be some more careful lifecycle tracking here. I don't think multiple handlers will be too easy, because the `#canvas-container` element will change when these need to change. But still, things get weird when handlers don't get cleaned up.

Also these should use the refs system like other elements.